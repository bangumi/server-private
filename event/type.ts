export const EventOp = Object.freeze({
  Create: 'c',
  Delete: 'd',
  Update: 'u',
  Snapshot: 'r',
} as const);

export type EventOp = (typeof EventOp)[keyof typeof EventOp];

export interface Payload {
  // before: object;
  // after: object;
  source: {
    table: string;
  };
  op: EventOp;
}

export interface KafkaMessage {
  topic: string;
  key: string;
  value: string;
}
