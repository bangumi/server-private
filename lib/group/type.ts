export enum GroupTopicMode {
  All = 'all',
  Joined = 'joined',
}

export enum GroupSort {
  Trends = 'trends',
  Created = 'created',
  Updated = 'updated',
}
