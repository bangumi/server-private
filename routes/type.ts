import type { IncomingMessage, ServerResponse } from 'node:http';

import type { Static, TSchema } from '@sinclair/typebox';
import type {
  ContextConfigDefault,
  FastifyBaseLogger,
  FastifyInstance,
  FastifyReply,
  FastifyRequest,
  FastifySchema,
  FastifyTypeProvider,
  RawReplyDefaultExpression,
  RawRequestDefaultExpression,
  RawServerDefault,
  RouteGenericInterface,
  RouteHandlerMethod,
} from 'fastify';
import type { ResolveFastifyRequestType } from 'fastify/types/type-provider.d.ts';

// from https://github.com/fastify/fastify-type-provider-typebox/blob/v3.1.0/index.ts#L56
export interface TypeBoxTypeProvider extends FastifyTypeProvider {
  output: this['input'] extends TSchema ? Static<this['input']> : never;
}

export type App = FastifyInstance<
  RawServerDefault,
  IncomingMessage,
  ServerResponse,
  FastifyBaseLogger,
  TypeBoxTypeProvider
>;

export interface Option {
  tags?: string[];
}

export type Request<schema extends FastifySchema> = FastifyRequest<
  RouteGenericInterface,
  RawServerDefault,
  RawRequestDefaultExpression,
  schema,
  TypeBoxTypeProvider,
  ContextConfigDefault,
  FastifyBaseLogger,
  ResolveFastifyRequestType<TypeBoxTypeProvider, schema, RouteGenericInterface>
>;

export type Reply<schema extends FastifySchema> = FastifyReply<
  RawServerDefault,
  RawRequestDefaultExpression,
  RawReplyDefaultExpression,
  RouteGenericInterface,
  ContextConfigDefault,
  schema,
  TypeBoxTypeProvider
>;

export type Handler<schema extends FastifySchema> = RouteHandlerMethod<
  RawServerDefault,
  RawRequestDefaultExpression,
  RawReplyDefaultExpression,
  RouteGenericInterface,
  ContextConfigDefault,
  schema,
  TypeBoxTypeProvider
>;
