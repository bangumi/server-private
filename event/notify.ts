import { fromBinary } from '@bufbuild/protobuf';
import { DateTime } from 'luxon';

import { db } from '@app/drizzle';
import { Notify, type NotifyType } from '@app/lib/notify.ts';
import { NotifySchema } from '@app/vendor/proto/mq/notify_pb';

import { type KafkaMessage } from './type';

export async function handle({ value }: KafkaMessage) {
  const payload = fromBinary(NotifySchema, Buffer.from(value, 'utf8'));

  await db.transaction(async (t) => {
    await Notify.create(t, {
      destUserID: payload.userId,
      sourceUserID: 0,
      createdAt: DateTime.now().toUnixInteger(),
      type: payload.type as NotifyType,
      relatedID: payload.mid,
      mainID: payload.mid,
      title: payload.title,
    });
  });
}
