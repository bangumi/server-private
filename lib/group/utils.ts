import { db, op, schema } from '@app/drizzle';

/** Is user member of group */
export async function isMemberInGroup(userID: number, groupID: number): Promise<boolean> {
  const [d] = await db
    .select({ uid: schema.chiiGroupMembers.uid, gid: schema.chiiGroupMembers.gid })
    .from(schema.chiiGroupMembers)
    .where(
      op.and(
        op.eq(schema.chiiGroupMembers.uid, userID),
        op.eq(schema.chiiGroupMembers.gid, groupID),
      ),
    )
    .limit(1);

  return Boolean(d);
}
