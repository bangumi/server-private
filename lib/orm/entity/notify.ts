import { Column, Entity, Index, PrimaryGeneratedColumn } from 'typeorm';

import type { Transformer } from '@app/lib/orm/transformer.ts';

@Index('nt_from_uid', ['from_uid'], {})
@Index('nt_mid', ['notify_field_id'], {})
@Index('nt_uid', ['uid', 'unread', 'type', 'postID'], {})
@Entity('chii_notify', { schema: 'bangumi' })
export class Notify {
  @PrimaryGeneratedColumn({ type: 'mediumint', name: 'nt_id', unsigned: true })
  id!: number;

  @Column('mediumint', { name: 'nt_uid', unsigned: true })
  uid!: number;

  @Column('mediumint', { name: 'nt_from_uid', unsigned: true })
  from_uid!: number;

  @Column('tinyint', {
    name: 'nt_status',
    unsigned: true,
    default: () => "'1'",
    transformer: {
      to: (value: boolean) => (value ? 1 : 0),
      from: Boolean,
    } satisfies Transformer<number, boolean>,
  })
  unread!: boolean;

  @Column('tinyint', { name: 'nt_type', unsigned: true, default: () => "'0'" })
  type!: number;

  @Column('mediumint', {
    name: 'nt_mid',
    comment: 'ID in notify_field',
    unsigned: true,
  })
  notify_field_id!: number;

  @Column('int', { name: 'nt_related_id', unsigned: true })
  postID!: number;

  @Column('int', { name: 'nt_dateline', unsigned: true })
  dateline!: number;
}

@Index('ntf_hash', ['hash'], {})
@Index('ntf_rid', ['topicID'], {})
@Entity('chii_notify_field', { schema: 'bangumi' })
export class NotifyField {
  @PrimaryGeneratedColumn({ type: 'mediumint', name: 'ntf_id', unsigned: true })
  id!: number;

  @Column('tinyint', { name: 'ntf_hash', unsigned: true, default: () => "'0'" })
  hash!: number;

  @Column('int', { name: 'ntf_rid', unsigned: true })
  topicID!: number;

  @Column('varchar', { name: 'ntf_title', length: 255 })
  title!: string;
}
