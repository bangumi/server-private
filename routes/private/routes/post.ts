import type { Static } from '@sinclair/typebox';
import { Type as t } from '@sinclair/typebox';
import { DateTime } from 'luxon';

import type { IAuth } from '@app/lib/auth';
import { NotAllowedError } from '@app/lib/auth';
import { Dam } from '@app/lib/dam';
import { BadRequestError, NotFoundError } from '@app/lib/error';
import * as Notify from '@app/lib/notify';
import { Security, Tag } from '@app/lib/openapi';
import type { IBaseReply } from '@app/lib/orm';
import { fetchUserX, GroupRepo } from '@app/lib/orm';
import * as orm from '@app/lib/orm';
import { CommentState, NotJoinPrivateGroupError } from '@app/lib/topic';
import * as Topic from '@app/lib/topic';
import { formatErrors, toResUser } from '@app/lib/types/res';
import * as res from '@app/lib/types/res';
import { requireLogin } from '@app/routes/hooks/pre-handler';
import type { App } from '@app/routes/type';

// eslint-disable-next-line @typescript-eslint/require-await
export async function setup(app: App) {
  app.addSchema(res.Error);
  app.addSchema(res.User);
  const BasicReply = t.Object(
    {
      id: t.Integer(),
      creator: t.Ref(res.User),
      createdAt: t.Integer(),
      text: t.String(),
      state: t.Integer(),
    },
    { $id: 'BasicReply' },
  );

  const Reply = t.Object(
    {
      ...BasicReply.properties,
      topicID: t.Integer(),
      topicTitle: t.String(),
    },
    { $id: 'GroupReply' },
  );

  app.addSchema(BasicReply);
  app.addSchema(Reply);

  async function getGroupPost(auth: IAuth, postID: number) {
    const post = await orm.GroupPostRepo.findOneBy({ id: postID });
    if (!post) {
      throw new NotFoundError(`group post ${postID}`);
    }

    if ([Topic.CommentState.UserDelete, Topic.CommentState.AdminDelete].includes(post.state)) {
      throw new NotFoundError(`group post ${postID}`);
    }

    const topic = await Topic.fetchDetail(auth, 'group', post.topicID);
    if (!topic) {
      throw new NotFoundError(`group topic ${post.topicID}`);
    }

    if (topic.contentPost.id === post.id) {
      throw new NotFoundError(`group post ${postID}`);
    }

    if ([Topic.CommentState.UserDelete, Topic.CommentState.AdminDelete].includes(topic.state)) {
      throw new NotFoundError(`group topic ${post.topicID}`);
    }

    return { post, topic };
  }

  app.get(
    '/groups/-/posts/:postID',
    {
      schema: {
        operationId: 'getGroupPost',
        params: t.Object({
          postID: t.Integer({ examples: [2092074] }),
        }),
        tags: [Tag.Group],
        response: {
          200: t.Ref(Reply),
          404: t.Ref(res.Error, {
            'x-examples': formatErrors(NotFoundError('post')),
          }),
        },
        security: [{ [Security.CookiesSession]: [] }],
      },
    },
    async ({ auth, params: { postID } }): Promise<Static<typeof Reply>> => {
      const { topic, post } = await getGroupPost(auth, postID);

      const creator = res.toResUser(await fetchUserX(post.uid));

      return {
        id: postID,
        creator,
        topicID: topic.id,
        state: post.state,
        createdAt: post.dateline,
        text: post.content,
        topicTitle: topic.title,
      };
    },
  );

  app.put(
    '/groups/-/posts/:postID',
    {
      schema: {
        operationId: 'editGroupPost',
        params: t.Object({
          postID: t.Integer({ examples: [2092074] }),
        }),
        tags: [Tag.Group],
        response: {
          200: t.Object({}),
          401: t.Ref(res.Error, {
            'x-examples': formatErrors(NotAllowedError('edit reply')),
          }),
        },
        security: [{ [Security.CookiesSession]: [] }],
        body: t.Object(
          {
            text: t.String({ minLength: 1 }),
          },
          {
            examples: [{ text: 'new post contents' }],
          },
        ),
      },
      preHandler: [requireLogin('edit a post')],
    },
    /**
     * @param auth -
     * @param text - 回帖内容
     * @param postID - 回复 ID
     */
    async function ({ auth, body: { text }, params: { postID } }): Promise<Record<string, never>> {
      if (auth.permission.ban_post) {
        throw new NotAllowedError('create reply');
      }

      if (!Dam.allCharacterPrintable(text)) {
        throw new BadRequestError('text contains invalid invisible character');
      }

      const post = await orm.GroupPostRepo.findOneBy({ id: postID });
      if (!post) {
        throw new NotFoundError(`post ${postID}`);
      }

      if (post.uid !== auth.userID) {
        throw new NotAllowedError('edit reply not created by you');
      }

      const topic = await Topic.fetchDetail(auth, 'group', post.topicID);
      if (!topic) {
        throw new NotFoundError(`topic ${post.topicID}`);
      }

      if (topic.state === CommentState.AdminCloseTopic) {
        throw new NotAllowedError('edit reply in a closed topic');
      }

      if ([CommentState.AdminDelete, CommentState.UserDelete].includes(topic.state)) {
        throw new NotAllowedError('edit a deleted reply');
      }

      for (const reply of topic.replies) {
        if (reply.id === post.id && reply.replies.length > 0) {
          throw new NotAllowedError('edit a reply with sub-reply');
        }
      }

      await orm.GroupPostRepo.update({ id: postID }, { content: text });

      return {};
    },
  );

  app.delete(
    '/groups/-/posts/:postID',
    {
      schema: {
        operationId: 'deleteGroupPost',
        params: t.Object({
          postID: t.Integer({ examples: [2092074] }),
        }),
        tags: [Tag.Group],
        response: {
          200: t.Object({}),
          401: t.Ref(res.Error, {
            'x-examples': formatErrors(NotAllowedError('delete this post')),
          }),
          404: t.Ref(res.Error, {
            'x-examples': formatErrors(NotFoundError('post')),
          }),
        },
        security: [{ [Security.CookiesSession]: [] }],
      },
      preHandler: [requireLogin('delete a post')],
    },
    async ({ auth, params: { postID } }) => {
      const { post } = await getGroupPost(auth, postID);

      if (auth.userID !== post.uid) {
        throw new NotAllowedError('delete this post');
      }

      if (post.state !== Topic.CommentState.Normal) {
        return {};
      }

      await orm.GroupPostRepo.update({ id: postID }, { state: Topic.CommentState.UserDelete });
      return {};
    },
  );

  app.post(
    '/groups/-/topics/:topicID/replies',
    {
      schema: {
        operationId: 'createGroupReply',
        params: t.Object({
          topicID: t.Integer({ examples: [371602] }),
        }),
        tags: [Tag.Group],
        response: {
          200: t.Ref(BasicReply),
          401: t.Ref(res.Error, {
            'x-examples': formatErrors(NotJoinPrivateGroupError('沙盒')),
          }),
        },
        security: [{ [Security.CookiesSession]: [] }],
        body: t.Object(
          {
            replyTo: t.Optional(
              t.Integer({
                examples: [0],
                default: 0,
                description: '被回复的 topic ID, `0` 代表回复楼主',
              }),
            ),
            content: t.String({ minLength: 1 }),
          },
          {
            examples: [
              { content: 'post contents' },
              {
                content: 'post contents',
                replyTo: 2,
              },
            ],
          },
        ),
      },
      preHandler: [requireLogin('creating a reply')],
    },
    /**
     * @param auth -
     * @param content - 回帖内容
     * @param relatedID - 子回复时的父回复ID，默认为 `0` 代表回复帖子
     * @param topicID - 帖子 ID
     */
    async ({
      auth,
      body: { content, replyTo = 0 },
      params: { topicID },
    }): Promise<Static<typeof BasicReply>> => {
      if (auth.permission.ban_post) {
        throw new NotAllowedError('create reply');
      }

      const topic = await Topic.fetchDetail(auth, 'group', topicID);
      if (!topic) {
        throw new NotFoundError(`topic ${topicID}`);
      }
      if (topic.state === CommentState.AdminCloseTopic) {
        throw new NotAllowedError('reply to a closed topic');
      }

      const now = DateTime.now();

      let parentID = 0;
      let dstUserID = topic.creatorID;
      if (replyTo) {
        const parents: Record<number, IBaseReply> = Object.fromEntries(
          topic.replies.flatMap((x): [number, IBaseReply][] => {
            // 管理员操作不能回复
            if (
              [
                CommentState.AdminCloseTopic,
                CommentState.AdminReopen,
                CommentState.AdminSilentTopic,
              ].includes(x.state)
            ) {
              return [];
            }
            return [[x.id, x], ...x.replies.map((x): [number, IBaseReply] => [x.id, x])];
          }),
        );

        const replied = parents[replyTo];

        if (!replied) {
          throw new NotFoundError(`parent post id ${replyTo}`);
        }

        dstUserID = replied.creatorID;
        parentID = replied.repliedTo || replied.id;
      }

      const group = await GroupRepo.findOneOrFail({
        where: { id: topic.parentID },
      });

      if (!group.accessible && !(await orm.isMemberInGroup(group.id, auth.userID))) {
        throw new NotJoinPrivateGroupError(group.name);
      }

      const t = await Topic.createTopicReply({
        topicType: Topic.Type.group,
        topicID: topicID,
        userID: auth.userID,
        content,
        parentID,
      });

      const notifyType = replyTo === 0 ? Notify.Type.GroupTopicReply : Notify.Type.GroupPostReply;
      await Notify.create({
        destUserID: dstUserID,
        sourceUserID: auth.userID,
        now,
        type: notifyType,
        postID: t.id,
        topicID: topic.id,
        title: topic.title,
      });

      return {
        id: t.id,
        state: t.state,
        createdAt: t.createdAt,
        text: t.content,
        creator: toResUser(t.user),
      };
    },
  );
}
