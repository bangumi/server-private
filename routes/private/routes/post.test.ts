import { DateTime } from 'luxon';
import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';

import type { IAuth } from '@app/lib/auth/index.ts';
import { emptyAuth, UserGroup } from '@app/lib/auth/index.ts';
import * as Notify from '@app/lib/notify.ts';
import * as orm from '@app/lib/orm/index.ts';
import type { ITopicDetails } from '@app/lib/topic/index.ts';
import { CommentState, TopicDisplay } from '@app/lib/topic/index.ts';
import * as Topic from '@app/lib/topic/index.ts';
import { createTestServer } from '@app/tests/utils.ts';

import { setup } from './post.ts';

/**
 * Topic 375793
 *
 * Reply 2177419 (287622)
 *
 * - Sub-reply 2177420 (287622)
 */

beforeEach(async () => {
  await orm.GroupPostRepo.update(
    {
      id: 2177420,
    },
    {
      state: 0,
      content: 'before-test',
    },
  );
});

describe('get post', () => {
  test('ok', async () => {
    const app = createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: 287622,
      },
    });

    await app.register(setup);

    const res = await app.inject({ method: 'get', url: '/groups/-/posts/2092074' });
    expect(res.json()).toMatchInlineSnapshot(`
      Object {
        "createdAt": 1662283112,
        "creator": Object {
          "avatar": Object {
            "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
            "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
            "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
          },
          "id": 287622,
          "nickname": "nickname 287622",
          "sign": "sing 287622",
          "user_group": 0,
          "username": "287622",
        },
        "id": 2092074,
        "state": 0,
        "text": "sub",
        "topicID": 371602,
        "topicTitle": "tes",
      }
    `);
  });

  test('not found', async () => {
    const app = createTestServer({});
    await app.register(setup);

    const res = await app.inject({ method: 'get', url: '/groups/-/posts/209207400' });
    expect(res.statusCode).toBe(404);
  });
});

describe('edit post', () => {
  test('should edit post', async () => {
    const app = createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: 287622,
      },
    });

    await app.register(setup);

    const res = await app.inject({
      url: '/groups/-/posts/2177420',
      method: 'put',
      payload: { text: 'new content' },
    });

    expect(res.statusCode).toBe(200);

    const pst = await orm.GroupPostRepo.findOneBy({
      id: 2177420,
    });

    expect(pst?.content).toBe('new content');
  });

  test('should not edit post', async () => {
    const app = createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: 287622 + 1,
      },
    });

    await app.register(setup);

    const res = await app.inject({
      url: '/groups/-/posts/2177420',
      method: 'put',
      payload: { text: 'new content' },
    });

    expect(res.json()).toMatchInlineSnapshot(`
      Object {
        "code": "NOT_ALLOWED",
        "error": "Unauthorized",
        "message": "you don't have permission to edit reply not created by you",
        "statusCode": 401,
      }
    `);
    expect(res.statusCode).toBe(401);
  });

  test('should not edit post with sub-reply', async () => {
    const app = createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: 287622,
      },
    });

    await app.register(setup);

    const res = await app.inject({
      url: '/groups/-/posts/2177419',
      method: 'put',
      payload: { text: 'new content' },
    });

    expect(res.json()).toMatchInlineSnapshot(`
      Object {
        "code": "NOT_ALLOWED",
        "error": "Unauthorized",
        "message": "you don't have permission to edit a reply with sub-reply",
        "statusCode": 401,
      }
    `);
    expect(res.statusCode).toBe(401);
  });
});

describe('create group post reply', () => {
  const createTopicReply = vi.fn().mockResolvedValue({
    id: 6,
    content: '',
    state: CommentState.Normal,
    createdAt: DateTime.fromISO('2021-10-21').toUnixInteger(),
    type: Topic.Type.group,
    topicID: 371602,
    user: {
      img: '',
      username: 'u',
      groupID: UserGroup.Normal,
      id: 9,
      nickname: 'n',
      regTime: DateTime.fromISO('2008-10-01').toUnixInteger(),
      sign: '',
    },
  });

  const notifyMock = vi.fn();
  beforeEach(() => {
    vi.spyOn(Topic, 'createTopicReply').mockImplementation(createTopicReply);
    vi.spyOn(Notify, 'create').mockImplementation(notifyMock);
    vi.spyOn(Topic, 'fetchDetail').mockImplementationOnce(
      (_: IAuth, type: 'group', id: number): Promise<ITopicDetails | null> => {
        if (id !== 371602) {
          return Promise.resolve(null);
        }

        return Promise.resolve({
          contentPost: { id: 100 },
          replies: [],
          creatorID: 287622,
          id: id,
          title: 't',
          display: TopicDisplay.Normal,
          createdAt: DateTime.now().toUnixInteger(),
          text: 't',
          state: CommentState.Normal,
          parentID: 1,
        });
      },
    );
  });

  afterEach(() => {
    vi.resetModules();
  });

  test('should create group post reply', async () => {
    const app = createTestServer({
      auth: {
        groupID: UserGroup.Normal,
        login: true,
        permission: {},
        allowNsfw: true,
        regTime: 0,
        userID: 100,
      },
    });

    await app.register(setup);

    const res = await app.inject({
      url: '/groups/-/topics/371602/replies',
      method: 'post',
      payload: {
        content: 'post contents',
      },
    });

    expect(res.json()).toMatchObject({
      creator: {
        avatar: {
          large: 'https://lain.bgm.tv/pic/user/l/icon.jpg',
          medium: 'https://lain.bgm.tv/pic/user/m/icon.jpg',
          small: 'https://lain.bgm.tv/pic/user/s/icon.jpg',
        },
        id: 9,
        nickname: 'n',
        sign: '',
        user_group: 10,
        username: 'u',
      },
      id: 6,
      state: 0,
      text: '',
    });
    expect(res.statusCode).toBe(200);
    expect(notifyMock).toHaveBeenCalledOnce();
    expect(notifyMock).toBeCalledWith(
      expect.objectContaining({
        destUserID: 287622,
        type: Notify.Type.GroupTopicReply,
      }),
    );
  });

  test('should not create with banned user', async () => {
    const app = createTestServer({
      auth: {
        groupID: UserGroup.Normal,
        login: true,
        permission: {
          ban_post: true,
        },
        regTime: 0,
        allowNsfw: true,
        userID: 1,
      },
    });

    await app.register(setup);

    const res = await app.inject({
      url: '/groups/-/topics/371602/replies',
      method: 'post',
      payload: {
        content: 'post contents',
      },
    });

    expect(res.statusCode).toBe(401);
    expect(createTopicReply).toBeCalledTimes(1);
  });

  test('should not create on non-existing topic', async () => {
    const app = createTestServer({
      auth: {
        groupID: UserGroup.Normal,
        login: true,
        permission: {},
        allowNsfw: true,
        regTime: 0,
        userID: 1,
      },
    });

    await app.register(setup);

    const res = await app.inject({
      url: '/groups/-/topics/3716000/replies',
      method: 'post',
      payload: {
        content: 'post contents',
      },
    });

    expect(res.statusCode).toBe(404);
    expect(res.json()).toMatchSnapshot();
  });

  test('should not create on non-existing topic reply', async () => {
    const app = createTestServer({
      auth: {
        groupID: UserGroup.Normal,
        login: true,
        permission: {},
        allowNsfw: true,
        regTime: 0,
        userID: 1,
      },
    });

    await app.register(setup);

    const res = await app.inject({
      url: '/groups/-/topics/371602/replies',
      method: 'post',
      payload: {
        content: 'post contents',
        replyTo: 11,
      },
    });

    expect(res.statusCode).toBe(404);
    expect(res.json()).toMatchSnapshot();
  });
});

async function testServer(...arg: Parameters<typeof createTestServer>) {
  const app = createTestServer(...arg);

  await app.register(setup);
  return app;
}

describe('delete post', () => {
  beforeEach(() => {
    vi.restoreAllMocks();
  });

  test('not found', async () => {
    const app = await testServer({ auth: { login: true, userID: 1 } });
    const res = await app.inject({
      url: '/groups/-/posts/2092074123',
      method: 'delete',
    });

    expect(res.statusCode).toBe(404);
  });

  test('ok', async () => {
    const app = await testServer({ auth: { login: true, userID: 287622 } });

    const res = await app.inject({ method: 'delete', url: '/groups/-/posts/2177420' });
    expect(res.json()).toMatchSnapshot();
    expect(res.statusCode).toBe(200);
  });

  test('not allowed not login', async () => {
    const app = await testServer();
    const res = await app.inject({ url: '/groups/-/posts/2177420', method: 'delete' });

    expect(res.json()).toMatchSnapshot();
    expect(res.statusCode).toBe(401);
  });

  test('not allowed wrong user', async () => {
    const app = await testServer({ auth: { login: true, userID: 1122 } });
    const res = await app.inject({ url: '/groups/-/posts/2177420', method: 'delete' });

    expect(res.json()).toMatchSnapshot();
    expect(res.statusCode).toBe(401);
  });
});
