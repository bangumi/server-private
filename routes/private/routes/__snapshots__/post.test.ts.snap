// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`create group post reply > should not create on non-existing topic 1`] = `
Object {
  "code": "NOT_FOUND",
  "error": "Not Found",
  "message": "topic 3716000 not found",
  "statusCode": 404,
}
`;

exports[`create group post reply > should not create on non-existing topic reply 1`] = `
Object {
  "code": "NOT_FOUND",
  "error": "Not Found",
  "message": "parent post id 11 not found",
  "statusCode": 404,
}
`;

exports[`delete group post > not allowed not login 1`] = `
Object {
  "code": "NEED_LOGIN",
  "error": "Unauthorized",
  "message": "you need to login before delete a post",
  "statusCode": 401,
}
`;

exports[`delete group post > not allowed wrong user 1`] = `
Object {
  "code": "NOT_ALLOWED",
  "error": "Unauthorized",
  "message": "you don't have permission to delete this post",
  "statusCode": 401,
}
`;

exports[`delete group post > ok 1`] = `Object {}`;

exports[`delete subject post > not allowed not login 1`] = `
Object {
  "code": "NEED_LOGIN",
  "error": "Unauthorized",
  "message": "you need to login before delete a post",
  "statusCode": 401,
}
`;

exports[`delete subject post > not allowed wrong user 1`] = `
Object {
  "code": "NOT_ALLOWED",
  "error": "Unauthorized",
  "message": "you don't have permission to delete this post",
  "statusCode": 401,
}
`;

exports[`delete subject post > ok 1`] = `Object {}`;
